src/pytorch_lightning/utilities/auto_restart.py

```diff
@@ -601,10 +601,11 @@ def _rotate_worker_indices(state: Dict[int, Any], latest_worker_id: int, num_wor
     on."""
     if num_workers == 0:
         return state
+    # the state of the first worker is the same as the state of all workers.
+    if len(state) == 1:
+        state = {i: state[0] for i in range(num_workers)}
```

Comment:
> This fix assumes that the state of all DataLoader workers is the same, which should be the case following my understanding, but please correct me if I'm wrong.

So, the state can be different in each worker because each worker usually gets a different seed. And you can see that e.g. on line 255 we collect the random state in the getitem, which runs inside each worker process.

This is my quick immediate answer here, but I need to spend more time now studying your fix (and the actual problem)